src/controllers/api-key.js

```diff
@@ -10,52 +10,184 @@
+      }
+    });
+
+    if (!hasText(data.name)) {
+      throw new ErrorWithStatusCode('Invalid request: missing name in the request data', STATUS_BAD_REQUEST);
+    }
+  }
+
+  /**
+   * Validate the IMS Org ID.
+   * @param imsOrgId
+   * @param imsUserToken
+   */
+  async function validateImsOrgId(imsOrgId, imsUserToken) {
+    if (!imsOrgId) {
+      throw new ErrorWithStatusCode('Missing x-ims-gw-org-id header', STATUS_UNAUTHORIZED);
+    }
+    const imsUserProfile = await imsClient.getImsUserProfile(imsUserToken);
+    log.debug(`Retrieved the IMS User Profile: ${JSON.stringify(imsUserProfile)}`);
+    const { organizations } = imsUserProfile;
+    if (!organizations.includes(imsOrgId)) {
+      throw new ErrorWithStatusCode('Invalid request: Organization not found', STATUS_UNAUTHORIZED);
+    }
   }
 
-  // eslint-disable-next-line no-unused-vars
-  function validateImsOrgId(imsOrgId) {
-    // Not implemented yet
+  /**
+   * Get the IMS user token from the headers.
+   * @param headers
+   * @returns string imsUserToken
+   */
+  function getImsUserToken(headers) {
+    const authorizationHeader = headers.authorization;
+    if (!authorizationHeader) {
+      throw new ErrorWithStatusCode('Missing Authorization header', STATUS_UNAUTHORIZED);
+    }
+    return authorizationHeader.replace('Bearer ', '');
   }
 
   /**
    * Create a new API key.
    * @param {Object} requestContext - Context of the request.
-   * @returns {Promise<Response>} 201 Created with the new API key.
+   * @returns {Promise<ApiKey>} 201 Created with the new API key.
    */
   async function createApiKey(requestContext) {
     const { data, pathInfo: { headers } } = requestContext;
     const imsOrgId = headers['x-ims-gw-org-id'];
 
-    validateRequestData(data);
-    validateImsOrgId(imsOrgId);
+    try {
+      const imsUserToken = getImsUserToken(headers);
+
+      validateRequestData(data);
+      await validateImsOrgId(imsOrgId, imsUserToken);
+
+      // Check if the domains are within the limit
+      if (data.domains.length > maxDomainsPerApiKey) {
+        throw new ErrorWithStatusCode('Invalid request: Exceeds the number of domains allowed', STATUS_FORBIDDEN);
+      }
+
+      const { authInfo: { profile } } = attributes;
+
+      // Currently the email is assigned as the imsUserId
+      const imsUserId = profile.email;
+
+      // Check whether the user has already created the maximum number of
+      // active API keys for the given imsOrgId.
+      const apiKeys = await dataAccess.getApiKeysByImsUserIdAndImsOrgId(imsUserId, imsOrgId);
+
+      log.debug('Retrieved the API keys for the user: ', apiKeys);
+
+      const validApiKeys = apiKeys && apiKeys.filter(
+        (apiKey) => apiKey.isValid(),
+      );
+
+      if (validApiKeys && validApiKeys.length >= maxApiKeys) {
+        throw new ErrorWithStatusCode('Invalid request: Exceeds the number of API keys allowed', STATUS_FORBIDDEN);
```

Comment:
Please include the maximum number of keys in this message
```suggestion
        throw new ErrorWithStatusCode(`Invalid request: You have exceeded the maximum number of API keys allowed (${maxApiKeys})`, STATUS_FORBIDDEN);
```